feat(new auth api): add EmbyConnect authentication API

pXius opened the pull request:
### Description
This PR is to serve as an RFC to add EmbyConnect authentication to Jellyseerr.

If follows the exact flow used when login in via https://emby.media, but creates a `JellyfinLoginResponse` as the last step using the acquired data.

```mermaid
graph TD
    A{Start flow IF username <br> is an email address <br> && <br> Jellyfin auth failed} --> B[Authenticate towards EmbyConnect <br> with email and password]
    B --> C[Get a list of Emby servers linked to user]
    C --> D{Compare if a linked <br> server matches the Emby <br> instance in Jellyseerr config}
    D -->|No Match| F[Throw No matching server Exception]
    D -->|Match Found| E[Do an authentication exchange with the hosted <br> Emby server to get a local userId and authToken]
    E --> G[Construct a JellyfinLoginResponse object <br> and return to existing Jellyfin auth login flow]
```

### Changes to Code

- Created an emby connect api with the required methods to orchestrate a set of calls towards EmbyConnect to build a `JellyseerrLoginResponse` object
- Add an EmbyConnect auth call to the Jellyfin login() method as the last auth attempt _IF_ the username is an email address
- Update the post() method on the ExternalApi class to handle request bodies that need to be `x-www-form-urlencoded`

### Notes
- This flow will NOT work as first time setup login as there is no existing configured Emby server to match to. First auth has to use a local account as it is today.
- I attempted to encapsulate as much of the logic in one place to avoid any possible conflicts with #685 and other future in-progress work.
- _JS is not my strong suit; **any** comments for improvement are welcome._

### Additional Improvements?
- Check new server type from #685 before allowing flow. This would avoid calls to EmbyConnect if someone has Jellyfin users with email addresses as names. 

#### To-Dos

- [X] Successful build `pnpm build`
- [X] Translation keys `pnpm i18n:extract`
- [ ] Database migration (if required)

#### Issues Fixed or Closed

- Fixes #749


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Emby Connect authentication supported for compatible servers.
  * Login form accepts email addresses (placeholder updated: "Email Address / Username").

* **Behavior Improvements**
  * Emby Connect users can be linked by email to existing accounts to avoid duplicate accounts; avatars and usernames preserved/updated where available.
  * Emby Connect failures gracefully fall back to local authentication.

* **Chores**
  * Added mise.toml to .gitignore.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pXius committed:
feat(new auth api): add EmbyConnect authentication API

Created an Emby Connect API with the required methods to orchestrate a set of calls towards
EmbyConnect to build a JellyseerrLoginResponse object. Added an EmbyConnect authentication call to
the Jellyfin login() method as the last authentication attempt, but only if the username is an email
address. Updated the post() method on the ExternalApi class to handle request bodies that need to be
x-www-form-urlencoded.

#749

gauthier-th commented on server/api/embyconnect.ts:
```diff
@@ -0,0 +1,207 @@
+
+  public async authenticateConnectUser(Email?: string, Password?: string) {
+    logger.debug(
+      `Attempting to authenticate via EmbyConnect with email: ${Email}`
```
You can move the email inside of an object in the 2nd parameter of the logger.debug method to avoid the security warning from GitHub (like `logger.debug('...', { email })`)

pXius replied:
Done

gauthier-th commented on server/api/embyconnect.ts:
```diff
@@ -0,0 +1,207 @@
+  }
+}
+
+class EmbyServerApi extends ExternalAPI {
```
Could you move the only method of this class to the EmbyConnectAPI class? I don't think it's a good idea to create a class just for one method.

pXius replied:
Done.

The reason for the additional class was to bypass the baseUrl already passed to the EmbyConnectAPI constructor, because this call requires the server host url and not the EmbyConnect address.

In order to satisfy the request to only have a method, I had to expose the `overwriteBaseUrl?: string` argument in the ExternalApi `get()` method. I don't believe this is an issue as it's already done in other methods.

gauthier-th replied:
Ok I get it. In my opinion it makes more sense to not use this overwriteBaseUrl option. This option was introduced while transitionning from Axios to the Fetch API, for compatibility reason.
I think it makes things confusing to use this, since the `.get`, `.post`, etc methods should be used to GET/POST/etc data only on the same domain used everywhere in the class.
You could use the regular Fetch API instead, with something like this:
```js
const res = await fetch('...', { headers: { ... }});
if (!res.ok) throw new Error('...');
return await res.json();
```

pXius committed:
refactor: resolving PR commets

This refactor includes a change that adds a conditional arg to the ExternalApi get() method to
override the base url. This method was pre-existing and already used in the calls and method.

#943

pXius commented on server/api/jellyfin.ts:
```diff
@@ -169,11 +173,36 @@ class JellyfinAPI extends ExternalAPI {
     }
   }
 
+  private async authenticateWithEmbyConnect(
```
Done.

pXius commented:
@gauthier-th 

Thanks for the feedback. 
I've resolved all comments as requested, please confirm that [this](https://github.com/Fallenbagel/jellyseerr/pull/943#discussion_r1732871331) solution is acceptable.

pXius committed:
refactor: resolving PR comments

Reverting get() baseUrl overwrite in favour of a vanilla fetch() call

#943

pXius commented:
@gauthier-th 
Thanks again for the guidance.

I have resolved the final comment by reverting the baseUrl override in favour of a vanilla `fetch()` call as requested.

Please let me know if any other changes are required.

pXius commented:
Hey @gauthier-th 
Anything outstanding to get this merged?

gauthier-th commented:
> Hey @gauthier-th Anything outstanding to get this merged?

This will be merged, in time.
We also need more people reviewing and testing this, as I don't use EmbyConnect.

gauthier-th reviewed (dismissed):
LGTM.
This still needs to be tested, though.

ysmilda commented:
As this is a feature I'd love to have, I installed it on my local instance. The steps I took are:

```bash
git clone https://github.com/pXius/jellyseerr.git
cd jellyseerr
git checkout feature-emby-connect-login
docker build . -t fallenbagel/jellyseerr:embyconnectfix -f Dockerfile.local
```

After changing the tag on my `docker-compose.yml` and a `docker compose up -d` it seems to work great! Going to keep the development version up for a while to see if I run into any issues down the line.

gauthier-th commented:
@pXius can you please fix the merge conflict?
I'd also like to have one or two more people testing this before we can merge.

ysmilda commented:
After a month of running I've had no issues. Would love to see this merged as using the dev server introduces some performance degradation.

ysmilda commented:
Initially tried to build a non dev version, but ran into problems with the version tag and the ui reloading because of it. 

Thanks for the build, will switch over to that one for further testing.

markschrik commented:
Is there any clue when or if this will be merged into main? For now I am fine running the docker tag, but it would be great to use LATEST.

gauthier-th commented:
We are waiting for @pXius to fix the merge conflict. Once it's done it should be merged within a couple weeks.

markschrik commented:
Hm, sad that I can't see the conflict without making a request myself. I would love to help.

Fallenbagel commented:
> Hm, sad that I can't see the conflict without making a request myself. I would love to help.

5 comments above
![Screenshot_20241129_141914_GitHub.jpg](https://github.com/user-attachments/assets/a0bdff44-6473-4dc8-9c65-d9480569d726)

markschrik commented:
I know but I cannot see the actual conflict because I am nota member of this PR.

markschrik commented:
For now I am testing the docker tag, it works good :)

Updated the DB myself to convert my existing local accounts to work with Emby accounts, just had to add the JellyfinUsername and JellyfinUserId to the db and it worked.

pXius commented:
Hey guys, apologies for a late response. I was away for work. 

I'll pull in main and fix conflicts this weekend. 👌

markschrik commented:
> Hey guys, apologies for a late response. I was away for work.
> 
> I'll pull in main and fix conflicts this weekend. 👌

Nice. That would be amazing. 

Can you please try to make it work for local accounts that already have Emby connect, but the email is identical?

right now, the behaviour is, a user logs in with Emby connect, an a sql error gets logged for duplicate user.

markschrik commented:
Hey @pXius, Can you fix the conflicts please?

pXius committed:
feat: merging develop into #943

pXius committed:
feat: fixing merge conflicts, adding user linking for emby connect

#943

pXius committed:
Merge branch 'develop' of github.com:fallenbagel/jellyseerr into feature-emby-connect-login

pXius committed:
feat(component and api): adding emby connect authentication with local account linking

Updated authentication flow to handle local account linking

#943

pXius committed:
feat(component and api): adding emby connect authentication with local account linking

Updated authentication flow to handle local account linking

#943

pXius commented:
Sorry for the delay in addressing this, #life.
As always suggestions and comments are welcome. 

**Changes:**

- Resolved merge conflicts.
- Updated the authentication flow to link a successful Emby Connect authentication to a user with the same email address.
- Updated the form placeholder text to indicate that an email can be used as a username when connecting to Emby.
- Refactored code to accommodate the new authentication flow.

**Behavior:**

- Users can log in with their Emby Connect account even if they don't have a Jellyseerr account—a new user will be created automatically.
- Users can log in with their local Emby username and then authenticate via Emby Connect. The accounts will link, and the email will be updated accordingly.
- Users can create a local Jellyseerr account with their Emby Connect email address and then authenticate via Emby Connect. The accounts will link, updating the username to match Emby.
- When a local account is converted to an Emby-linked account, users can still log in using the Jellyseerr login form if they choose. However, they must use their linked email while retaining their Jellyseerr password, as this flow does not authenticate against the Emby server or Emby Connect.

pXius committed:
feat(component and api): embyConnect Auth Flow

Adding ability for emby connect auth flow to link to local accounts

#943

gauthier-th commented:
I've just updated the preview tag: `preview-emby-connect`

pXius committed:
Merge branch 'develop' of github.com:fallenbagel/jellyseerr into feature-emby-connect-login

pXius committed:
refactor: changing deviceId prefix from overseerr to jellyseerr

pXius committed:
Merge branch 'develop' of github.com:fallenbagel/jellyseerr into feature-emby-connect-login

pXius commented:
Hey everyone, 
Testing looks good on my side. 
If all looks well, I'd appreciate if we can get this merged in to avoid more merge conflicts as the #develop branch progresses.

Thanks!

fallenbagel commented:
> Hey everyone, 
> Testing looks good on my side. 
> If all looks well, I'd appreciate if we can get this merged in to avoid more merge conflicts as the #develop branch progresses.
> 
> Thanks!

Sorry for the long wait. We will be looking into reviewing these next 😄

pXius committed:
Merge branch 'develop' of github.com:fallenbagel/jellyseerr into feature-emby-connect-login

gauthier-th commented on server/api/externalapi.ts:
```diff
@@ -111,7 +112,10 @@ class ExternalAPI {
     config?: RequestInit
   ): Promise<T> {
-    const headers = { ...this.defaultHeaders, ...config?.headers };
+    const headers = new Headers({
+      ...this.defaultHeaders,
+      ...(config?.headers || {}),
+    });
```
Why are you changing this?

gauthier-th commented on server/api/externalapi.ts:
```diff
@@ -124,12 +128,24 @@ class ExternalAPI {
       ...config,
       headers,
-      body: data ? JSON.stringify(data) : undefined,
+      body: body,
```
```suggestion
      body,
```

gauthier-th commented on src/components/Login/JellyfinLogin.tsx:
```diff
@@ -138,7 +139,9 @@ const JellyfinLogin: React.FC<JellyfinLoginProps> = ({
                         name="username"
                         type="text"
-                        placeholder={intl.formatMessage(messages.username)}
+                        placeholder={`${intl.formatMessage(
+                          messages.email
+                        )} / ${intl.formatMessage(messages.username)}`}
```
@fallenbagel do you think we just just display "Username / Email Address" for Emby users?
I don't want Jellyfin users to get the wrong login and use their email instead of their username.

fallenbagel replied:
We have a similar placeholder on local login, which already causes tons of confusion amongst users (though it's not as heavily used as jellyfin login, so we can let it be for now).

But this would create a lot of confusion. I don't think this change would be a good idea.

fallenbagel replied:
Thats fine ig. Though influx of confusion will come from those emby users not using emby connect

gauthier-th commented on server/routes/auth.ts:
```diff
@@ -450,34 +453,59 @@ authRoutes.post('/jellyfin', async (req, res, next) => {
-          jellyfinUsername: account.User.Name,
-        }
-      );
+      // Emby Connect user with unlinked local account?
```
What does that mean? Is this a TODO?

gauthier-th commented on server/routes/auth.ts:
```diff
@@ -411,25 +405,34 @@ authRoutes.post('/jellyfin', async (req, res, next) => {
       user.avatar = getUserAvatarUrl(user);
       user.jellyfinUsername = account.User.Name;
 
+      if (
+        account.User.Email !== undefined &&
+        user.email !== account.User.Email
+      ) {
+        user.email = account.User.Email;
+      }
```
This is going to replace every username of Emby usernames by their email? I'm not sure that a silent change like this of all emails is desirable.

pXius replied:
The thinking here was to replace the local account email address with that of the emby connect account. 
The reason was to avoid the full flow of calls identifying if a user has an emby connect account, and then filtering out and trying to match an emby server to that of the JellyS instance every time a user logs in. 
The above flow adds delay and is only triggered after local auth fails. 

Ideally a user should be able to choose emby connect as a login option, but even emby users are confused by connect vs non connect accounts. 

I'll leave the decision to you guys.

pXius replied:
The above is outlined in the last point in the post about behavior. 

https://github.com/fallenbagel/jellyseerr/pull/943#issuecomment-2676378347

gauthier-th reviewed (changes requested):
Just a few more comments.
Sorry again for the repeated conflicts. We'll merge this into the next minor release once we've finished reviewing.

gauthier-th commented:
@pXius do you mind if take over this PR?

pXius committed:
feat: emby Connect

pXius committed:
feat: add Emby Connect Authentication Flow

This flow allows users to authneticate to JS using Emby Connect login details.

#943

pXius committed:
chore: address pnpm lock diff

pXius committed:
Merge remote-tracking branch 'origin' into feature-emby-connect-login

pXius committed:
refactor: revert changes to lock file by local formatter

gauthier-th commented:
Just updated the preview tag: `preview-emby-connect`
Will be available in ~40min

gauthier-th commented on src/components/Login/JellyfinLogin.tsx:
```diff
@@ -125,7 +126,9 @@ const JellyfinLogin: React.FC<JellyfinLoginProps> = ({
                         name="username"
                         type="text"
-                        placeholder={intl.formatMessage(messages.username)}
+                        placeholder={`${intl.formatMessage(
+                          messages.email
+                        )} / ${intl.formatMessage(messages.username)}`}
```
Leaving only "username" in this placeholder for Jellyfin users so they don't get confused:
```suggestion
                        placeholder={
                          settings.currentSettings.mediaServerType ===
                          MediaServerType.JELLYFIN
                            ? intl.formatMessage(messages.username)
                            : `${intl.formatMessage(
                                messages.email
                              )} / ${intl.formatMessage(messages.username)}`
                        }
```

merry-koala commented:
Hopefully, once seerr enters stable, this feature can be added.

Commit:
chore: resolving merge conflicts with develop branch

Resolving merge conflicts with develop branch

Commit:
chore: replacing email validator with Validator

Commit:
Update .gitignore

M0NsTeRRR commented on server/api/embyconnect.ts:
```diff
@@ -0,0 +1,211 @@
+      {},
+      {
+        headers: {
+          'X-Application': `Jellyseerr/${getAppVersion()}`,
```
```suggestion
          'X-Application': `Seerr/${getAppVersion()}`,
```

M0NsTeRRR commented on server/api/embyconnect.ts:
```diff
@@ -0,0 +1,211 @@
+      const params = new URLSearchParams({
+        format: 'json',
+        ConnectUserId: userId,
+        'X-Emby-Client': 'Jellyseerr',
```
```suggestion
        'X-Emby-Client': 'Seerr',
```

M0NsTeRRR commented on server/api/embyconnect.ts:
```diff
@@ -0,0 +1,211 @@
+        'X-Emby-Client': 'Jellyseerr',
+        'X-Emby-Device-Id': deviceId ?? uniqueId(),
+        'X-Emby-Client-Version': getAppVersion(),
+        'X-Emby-Device-Name': 'Jellyseerr',
```
```suggestion
        'X-Emby-Device-Name': 'Seerr',
```

M0NsTeRRR commented on server/routes/auth.ts:
```diff
@@ -453,34 +461,57 @@ authRoutes.post('/jellyfin', async (req, res, next) => {
+        await userRepository.save(user);
+      } else {
+        logger.info(
+          'Sign-in attempt from Jellyfin/Emby user with access to the media server; creating new Jellyseerr user',
```
```suggestion
          'Sign-in attempt from Jellyfin/Emby user with access to the media server; creating new Seerr user',
```

M0NsTeRRR reviewed (changes requested):
Hello,
We gonna move forward with your PR sorry for the delay @pXius :)
Having both a username and an email improves user QoL 👍
